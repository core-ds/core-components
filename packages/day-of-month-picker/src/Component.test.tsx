import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';

import { DayOfMonthPicker } from './index';

describe('DayOfMonthPicker', () => {
    describe('Display tests', () => {
        it('should match snapshot', () => {
            expect(
                render(<DayOfMonthPicker onChange={() => {}} value={1} />).container,
            ).toMatchSnapshot();
        });
    });

    it('should set custom class', () => {
        const className = 'custom-class';
        const { container } = render(
            <DayOfMonthPicker onChange={() => {}} value={1} className={className} />,
        );

        expect(container.firstElementChild).toHaveClass(className);
    });

    describe('Open/Close behavior', () => {
        it('should open calendar on click', async () => {
            const { container } = render(<DayOfMonthPicker onChange={() => {}} value={1} />);
            const component = container.firstElementChild as HTMLDivElement;

            await waitFor(() => {
                fireEvent.click(component);
                expect(document.querySelector('.inputWrapper')).toBeInTheDocument();
            });
        });

        it('should open calendar on focus', async () => {
            const { container } = render(<DayOfMonthPicker onChange={() => {}} value={1} />);
            const component = container.firstElementChild as HTMLDivElement;

            await waitFor(() => {
                fireEvent.focus(component);
                expect(document.querySelector('.inputWrapper')).toBeInTheDocument();
            });
        });

        it('should close calendar on blur', async () => {
            const { container } = render(<DayOfMonthPicker onChange={() => {}} value={1} />);
            const component = container.firstElementChild as HTMLDivElement;

            await waitFor(() => {
                fireEvent.focus(component);
                expect(document.querySelector('.calendarContainer')).toBeInTheDocument();
            });

            await waitFor(() => {
                fireEvent.blur(component);
                expect(document.querySelector('.calendarContainer')).not.toBeInTheDocument();
            });
        });

        it('should close calendar on escape', async () => {
            const { container } = render(<DayOfMonthPicker onChange={() => {}} value={1} />);
            const component = container.firstElementChild as HTMLDivElement;

            await waitFor(() => {
                fireEvent.focus(component);
                expect(document.querySelector('.inputWrapper')).toBeInTheDocument();
            });

            fireEvent.keyDown(component, { key: 'Escape' });

            await waitFor(() => {
                expect(document.querySelector('.calendarContainer')).not.toBeInTheDocument();
            });
        });

        it('should toggle calendar on enter if input focused', async () => {
            const { container, queryByRole } = render(
                <DayOfMonthPicker onChange={() => {}} value={1} />,
            );
            const component = container.firstElementChild as HTMLDivElement;
            const input = queryByRole('textbox') as HTMLInputElement;

            await waitFor(() => {
                fireEvent.focus(component);
                expect(document.querySelector('.calendarContainer')).toBeInTheDocument();
            });

            fireEvent.keyDown(component, { key: 'Enter' });

            await waitFor(() => {
                expect(document.querySelector('.calendarContainer')).toBeInTheDocument();
            });

            fireEvent.keyDown(input, { key: 'Enter' });

            await waitFor(() => {
                expect(document.querySelector('.calendarContainer')).not.toBeInTheDocument();
            });

            fireEvent.keyDown(input, { key: 'Enter' });

            await waitFor(() => {
                expect(document.querySelector('.calendarContainer')).toBeInTheDocument();
            });
        });
    });

    describe('Controlled-way', () => {
        it('should set value to input', () => {
            const { queryByRole, rerender } = render(
                <DayOfMonthPicker value={1} onChange={() => {}} />,
            );

            expect(queryByRole('textbox')).toHaveValue('1');

            rerender(<DayOfMonthPicker value={31} onChange={() => {}} />);

            expect(queryByRole('textbox')).toHaveValue('В последний день месяца');
        });

        it('should set value to calendar', async () => {
            const { container, rerender, getByTestId } = render(
                <DayOfMonthPicker value={1} onChange={() => {}} />,
            );
            const component = container.firstElementChild as HTMLDivElement;
            await waitFor(() => {
                fireEvent.focus(component);
            });

            const selectedButton = document.querySelector(
                'td[aria-selected="true"]',
            ) as HTMLButtonElement;

            expect(selectedButton).toBeInTheDocument();

            const day = +(selectedButton.getAttribute('data-day') as string);

            expect(day).toBe(1);

            await waitFor(() => {
                rerender(<DayOfMonthPicker value={31} onChange={() => {}} />);
            });
            const component1 = container.firstElementChild as HTMLDivElement;
            await waitFor(() => {
                fireEvent.focus(component1);
            });
            expect(getByTestId('last-day-button')).toHaveTextContent('В последний день');
        });
    });

    describe('Callback tests', () => {
        it('should call onChange callback', async () => {
            const cb = jest.fn();
            const defaultDay = 1;
            const { getByText, container, getByTestId } = render(
                <DayOfMonthPicker value={defaultDay} onChange={cb} />,
            );
            const component = container.firstElementChild as HTMLDivElement;
            await waitFor(() => {
                fireEvent.focus(component);
            });
            getByText('3').click();
            const selectedDay = cb.mock.calls[0][0];

            expect(cb).toBeCalledTimes(1);
            expect(selectedDay).toBeTruthy();
            expect(selectedDay).toBe(3);
        });
    });

    describe('Render tests', () => {
        test('should unmount without errors', () => {
            const { unmount } = render(<DayOfMonthPicker value={31} onChange={() => {}} />);

            expect(unmount).not.toThrowError();
        });
    });
});
