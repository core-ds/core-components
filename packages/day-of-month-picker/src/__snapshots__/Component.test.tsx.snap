// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayOfMonthPicker Display tests should match snapshot 1`] = `
<div>
  <div
    role="button"
    tabindex="0"
  >
    <div
      class="component component m block hasRightAddons"
    >
      <div
        class="inner inner disabled filled hasInnerLabel"
      >
        <div
          class="inputWrapper"
        >
          <span
            aria-hidden="true"
            class="hiddenLabel"
          >
            Какого числа
          </span>
          <div
            class="label label"
          >
            <span
              class="labelInner"
            >
              Какого числа
            </span>
          </div>
          <div
            class="input"
          >
            <input
              aria-label="Какого числа"
              class="input input m hasInnerLabel"
              data-test-id="day-input-field"
              readonly=""
              type="text"
              value="1"
            />
          </div>
        </div>
        <div
          class="addons rightAddons"
        >
          <button
            class="component ghost s component ghost cc-icon-button primary component"
            data-test-id="icon"
            type="button"
          >
            <span
              class="text stretchText"
            >
              <span
                class="iconWrapper s center"
              >
                <svg
                  class="icon"
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M9 12v2H7v-2h2zm0 6v-2H7v2h2zm2-6h2v2h-2v-2zm2 4h-2v2h2v-2zm4-4v2h-2v-2h2z"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M6 2h2v2h8V2h2v2h3v16a3 3 0 01-3 3H6a3 3 0 01-3-3V4h3V2zm13 7H5v11a1 1 0 001 1h12a1 1 0 001-1V9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
